### PR TITLE
Don't emit any source files found under node_modules

### LINKED ISSUE #6964: Discovering .ts files under node_modules resulting in them being compiled
I uncovered this general issue while testing loading JavaScript files from `node_modules`, but realized (and confirmed) this applies to TypeScript files also.

Basically, when we search the `node_modules` folder, we [look for all supported extensions](https://github.com/Microsoft/TypeScript/blob/release-1.8/src/compiler/program.ts#L139) and add them to the compilation. Currently this means that if we find a TypeScript file in a Node module (e.g. someone published their package including the source), then we compile this also.

For example, if module `sausages` was published as follows (with the original source besides the compiler files):

```
 - node_modules
 |- sausages
   |- package.json
   |- core.ts
   |- core.js
```

Then we get the benefit of loading the `.ts` and getting the types from this. However on compiling an app which uses the `sausages/core` module, with `outDir: "out"`, I get the resulting output (notice how `core.js` from the `sausages` package has been compiled into a `node_modules` folder in my `outDir`).

screen shot 2016-02-08 at 9 42 56 am

I already added a flag to denote if a file was found by searching Node modules in [this pull request](https://github.com/Microsoft/TypeScript/pull/6928/files#diff-08a3cc4f1f9a51dbb468c2810f5229d3R72). I can refactor this to eliminate this issue if that seems reasonable.

### PR DESCRIPTION
**Pull request by billti**
This fixes #6964 to suppress emit for any source files located by searching in node_modules, not just JavaScript files.

### CODE REVIEW

File: src/compiler/program.ts [hunk 1]
```diff
@@ -1378,7 +1378,7 @@ namespace ts {
                 getSourceFile: program.getSourceFile,
                 getSourceFileByPath: program.getSourceFileByPath,
                 getSourceFiles: program.getSourceFiles,
-                getFilesFromNodeModules: () => jsFilesFoundSearchingNodeModules,
+                isSourceFileFromNodeModules: (file: SourceFile) => !!lookUp(sourceFilesFoundSearchingNodeModules, file.path),
```
Comment (mhegazy) [inline src/compiler/program.ts]: we have `isExternalLibraryImport`, so why not call this `isSourceFileFromExternalLibrary` ?

**Comment by mhegazy** [comment]
👍

**Comment by billti** [comment]
Added fix for #9427, which was caused by similar changes for JavaScript. Also changed the API per @mhegazy 's suggestion.

**Comment by billti** [comment]
Hmmm, must have missed a test file in a commit... all passed on my machine. Will take a look shortly.

File: tests/baselines/reference/project/nodeModulesMaxDepthIncreased/amd/nodeModulesMaxDepthIncreased.errors.txt [hunk 2]
```diff
@@ -1,4 +1,5 @@
-maxDepthIncreased/root.ts(4,1): error TS2322: Type 'string' is not assignable to type 'number'.
+maxDepthIncreased/root.ts(7,1): error TS2322: Type 'string' is not assignable to type 'number'.
+maxDepthIncreased/root.ts(8,13): error TS2339: Property 'test' does not exist on type 'typeof "C:/src/TypeScript/tests/cases/projects/NodeModulesSearch/maxDepthIncreased/node_modules/@...'.
```
Comment (billti) [inline tests/baselines/reference/project/nodeModulesMaxDepthIncreased/amd/nodeModulesMaxDepthIncreased.errors.txt]: This is causing the issue in the GitHub build... need to figure out where the full path is creeping in to the error message.

File: tests/baselines/reference/project/nodeModulesMaxDepthIncreased/amd/nodeModulesMaxDepthIncreased.errors.txt [hunk 3]
```diff
@@ -28,11 +29,22 @@ maxDepthIncreased/root.ts(4,1): error TS2322: Type 'string' is not assignable to
...
     m1.f1("test");
     m1.f2.a = 10;
-    m1.f2.person.age = "10"; // Error: Should be number
+    
+    m1.f2.person.age = "10";    // Should error if loaded the .js files correctly
     ~~~~~~~~~~~~~~~~
 !!! error TS2322: Type 'string' is not assignable to type 'number'.
+    let r1 = m4.test.charAt(2); // Should error if correctly not using the .js file but using @types info
+                ~~~~
+!!! error TS2339: Property 'test' does not exist on type 'typeof "C:/src/TypeScript/tests/cases/projects/NodeModulesSearch/maxDepthIncreased/node_modules/@...'.
```
Comment (billti) [inline tests/baselines/reference/project/nodeModulesMaxDepthIncreased/amd/nodeModulesMaxDepthIncreased.errors.txt]: @RyanCavanaugh I'm seeing this be the actual error (with full path) when a types package has an incorrect type. Is this expected? Doesn't seem very readable.

error

Comment (RyanCavanaugh) [inline tests/baselines/reference/project/nodeModulesMaxDepthIncreased/amd/nodeModulesMaxDepthIncreased.errors.txt]: I'm not sure how this happens. @vladima any ideas?

Comment (mhegazy) [inline tests/baselines/reference/project/nodeModulesMaxDepthIncreased/amd/nodeModulesMaxDepthIncreased.errors.txt]: file names are always absolute in the VS case. modules do not have a name themselves. so we will use the full file name. this is something we have always done. so not a new thing. we can change that to get a user friendly name for modules to use in errors.

Comment (billti) [inline tests/baselines/reference/project/nodeModulesMaxDepthIncreased/amd/nodeModulesMaxDepthIncreased.errors.txt]: Right, but even the error message itself it kinda weird and not that helpful `does not exist on type 'typeof "<full path to file>"`. You can't even see the end of the error it's so long. Would at least something like `does not exist on type of module <moduleid> from <filepath>` be possible?

Comment (mhegazy) [inline tests/baselines/reference/project/nodeModulesMaxDepthIncreased/amd/nodeModulesMaxDepthIncreased.errors.txt]: sure. but again you need a user-friendly version of `<module id>`, cause now it is the file path.

**Comment by billti** [comment]
I removed that one check causing the full path to be in the error for now. The subsequent line would be in error if the wrong file was referenced anyway, so this doesn't really lose any coverage. Hopefully the tests will be green on the CI server (eventually :weary: ) after this change.

**Pull request merged**